pkgs/stdenv/linux/default.nix

```diff
@@ -109,12 +106,13 @@ let
     '';
 
 
-  # The bootstrap process proceeds in several steps.
-
-
-  # Create a standard environment by downloading pre-built binaries of
-  # coreutils, GCC, etc.
-
+  # The bootstrap process proceeds in several steps. Our high-level plan
+  # is the following:
+  #
+  # - build binutils with bootstrapTools
+  # - build gcc with bootstrapTools
+  # - build glibc with bootstrapTools + gcc from nixpkgs (embed libgcc_s.so.1)
+  # - build gcc with nixpkgs toolchain
```

Comment:
Building gcc twice is going to slow down `staging` rebuilds... there is no parallelism at this point in the build graph because everything depends on stdenv.  @vcunat are you okay with this?

Reply:
I know, I've been following this.  It seems worth building gcc twice, as it should address quite a lot of issues.  The slow-down is mainly for development of stdenv-rebuilding stuff (and bisection); on Hydra there's way more weight on the total amount of work over _all_ packages and that won't change noticeably.

We can instead try making the bootstrap process a bit less slow than in the first version of this PR, partially inside this PR (or the same staging-next iteration).

Reply:
> on Hydra there's way more weight on the total amount of work over _all_ packages

Oh, I didn't mean it would require adding more machines to the Hydra pool.  The issue isn't the total number of CPU cycles for a rebuild stdenv.  The issue is the *total number of wall-clock hours needed to rebuild stdenv*, regardless of how many builders you have.

This PR adds several serial stages to the bootstrap (derivations that can't be built in parallel with anything else), and some of those derivations (cough cough `gettext`) are effectively single-threaded builds that don't even make meaningful use of multiple cores.

> It seems worth building gcc twice,

Except that [we don't need to do that](https://github.com/NixOS/nixpkgs/pull/208478#issuecomment-1368737116) to get any of the benefits...

Reply:
My point was that for Hydra the wallclock hours _for stdenv_ don't seem so important because it always needs to build also very many other packages after that.  (The total time to rebuild everything is long anyway and very often there's lots of work on another branch that can be done in parallel.)

I don't see rush.  I mean, the author did write upfront that it would probably take weeks to merge.  I perceived split stages as a further improvement of this, _if_ someone can do that, but maybe it's my lack of knowledge of gcc's build process.  Either way, I believe it would certainly be nice to have at least some solution to the issues within a couple months, because aarch64-linux is accumulating hacks in the meantime (old gcc version, recently [rustc](https://github.com/NixOS/nixpkgs/pull/209113)).

Reply:
> > It seems worth building gcc twice,
> Except that https://github.com/NixOS/nixpkgs/pull/208478#issuecomment-1368737116 to get any of the benefits...

I think you are arguing for an implementation detail. I would say the comment would not even change if we were to implement your suggestion.  You better use latest `gcc` to build `glibc`. You better use latest `glibc` to build `gcc`. Building `gcc` as is (this PR) or shredding it in parts by making changes in `gcc`'s build system (AFAIU that's the alternative you propose) does not change this high-level layout. Does it?

Reply:
I am absolute no expert on the bootstrap topic, but I am currently playing a bit around on compiling my system with x86-64-v3. As the boostrapTools version is v9 it does not know those architectures. When building on the default generic we may not require a second rebuild on GCC (as the gcc9 may already build a newer GCC which fits the flags), i one would require features of a newer GCC to build the final GCC, we would require a second GCC rebuild right? 

Building x86-64-vX is still a WIP and quite controversial discussion on that PR but on my opinion it would be nice to get it working and this PR does do a lot of nice changes (where I at least guess I understand more about bootstrap process) why I thought leaving the question here may be a good place

Reply:
> As the boostrapTools version is v9 it does not know those architectures

You mean `boostrapTools` is on `gcc-8` and it might not be enough to generate the code for `-v3`? It depends. If `gcc-8` supports `-mavx2`, `-mmovbe` and friends that might be good enough. Worth checking by trying to build a simple program.

But generally yes, you need to make sure not to pass unsupported flags to `gcc`.

Another note: it should be fine to mix code from a default ISA implied by `gcc` and fresh `v3` set that adds extra instructions on top. No need to do anything beyond the bootstrap .

Reply:
`gcc-8` seems to recognize at least those 2 I tried:

```
$ nix-shell -E 'with import <nixpkgs> {}; pkgs.mkShell.override { stdenv = pkgs.gcc8Stdenv; } {}'

$ gcc -v
Using built-in specs.
COLLECT_GCC=/nix/store/m8hkk78wvnqa0pr5162b0vx26rl5lvvw-gcc-8.5.0/bin/gcc
COLLECT_LTO_WRAPPER=/nix/store/m8hkk78wvnqa0pr5162b0vx26rl5lvvw-gcc-8.5.0/libexec/gcc/x86_64-unknown-linux-gnu/8.5.0/lto-wrapper
Target: x86_64-unknown-linux-gnu
Configured with: 
Thread model: posix
gcc version 8.5.0 (GCC) 

$ printf "int main() {}" | gcc -x c -c - -o /dev/null -mavx2 -mmovbe

$ echo $?
0
```

Reply:
Thanks for the feedback, I meant initially the arch flags (there is something like -march=x86-64-v3 on newer compiler), setting single flags would already work on the old GCC, that's right. Current I try to filter out those flags in the bootstrap process.
what my thought was, if the bootstraped GCC is kept without recompiling from a nixpkgs GCC, that should result in a different GCC when doing a recompile, as recompiling then is from a nixpkgs GCC (if I am right on that) and this was basically the question/input.

Reply:
Ah, I had no idea `-march=x86-64-v3` was a thing. Yeah, you either need to rebuild `gcc` again if you already filtered them out or use `CFLAGS_FOR_TARGET` (and friends) to only affect the final result. It is certainly using flags for final libraries. Worth double-checking if the build process uses the same for `gcc` executable itself.